### PR TITLE
Support metrics for IOPS and bandwidth real time

### PR DESCRIPTION
**Pull request by derekbit**
https://github.com/longhorn/longhorn/issues/2406

Depends on https://github.com/longhorn/longhorn-engine/pull/794 and https://github.com/longhorn/longhorn-instance-manager/pull/158

Signed-off-by: Derek Su [derek.su@suse.com](mailto:derek.su@suse.com)

### CODE REVIEW

**Comment by derekbit** [comment]
Case 1. Run `dd if=/dev/urandom of=30G bs=1M count=30000 oflag=direct` in a pod with a longhorn volume.

- IO status got from `iostat -cx 1 sda`
image

- IO status go from longhorn metrics
image


Case 2: Copy one 30GB file in the volume to the same volume.

image

File: metrics_collector/metrics_collector.go [hunk 1]
```diff
@@ -19,20 +19,22 @@ import (
 )
 
 func InitMetricsCollectorSystem(logger logrus.FieldLogger, currentNodeID string, ds *datastore.DataStore, kubeconfigPath string, proxyConnCounter util.Counter) {
+	logger.Info("Initialize metrics collector system")
```
Comment (innobead) [inline metrics_collector/metrics_collector.go:22]: `Initializing ....`

File: metrics_collector/volume_collector.go [hunk 2]
```diff
@@ -80,26 +160,63 @@ func (vc *VolumeCollector) Describe(ch chan<- *prometheus.Desc) {
...
+				engineClientProxy, err = vc.getEngineClientProxy(e)
+				if err == nil {
+					defer engineClientProxy.Close()
+
+					metrics, err = engineClientProxy.MetricsGet(e)
+					if err != nil {
+						vc.logger.WithError(err).Warnf("Failed to get metrics from volume %v from engine %v", e.Spec.VolumeName, e.Name)
+					}
+				} else {
+					vc.logger.WithError(err).Warnf("Failed to get engine for volume %v", v.Name)
```
Comment (innobead) [inline metrics_collector/volume_collector.go:191]: `Failed to get engine proxy of %v for volume %v ...`

File: metrics_collector/metrics_collector.go (hunk 1 above)
Comment (derekbit) [reply metrics_collector/metrics_collector.go]: Updated. Thank you.

File: metrics_collector/volume_collector.go (hunk 2 above)
Comment (derekbit) [reply metrics_collector/volume_collector.go]: Updated. Thank you.

File: engineapi/proxy_metrics.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import (
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+)
+
+func (p *Proxy) MetricsGet(e *longhorn.Engine) (*Metrics, error) {
+	metrics, err := p.grpcClient.MetricsGet(p.DirectToURL(e))
+	if err != nil {
+		return nil, err
+	}
+	return (*Metrics)(metrics), err
```
Comment (innobead) [inline engineapi/proxy_metrics.go:12]: the metrics returned from the grpc call is a metrics type definied in`github.com/longhorn/longhorn-instance-manager/pkg/client`, but here we define another Metrics type. They can be casted to each other w/o issues? Why not just use the type from instance manager?

File: metrics_collector/volume_collector.go [hunk 4]
```diff
@@ -80,26 +160,63 @@ func (vc *VolumeCollector) Describe(ch chan<- *prometheus.Desc) {
...
 }
 
+func (vc *VolumeCollector) getEngineClientProxy(engine *longhorn.Engine) (c engineapi.EngineClientProxy, err error) {
+	engineCliClient, err := controller.GetBinaryClientForEngine(engine, &engineapi.EngineCollection{}, engine.Status.CurrentImage)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get binary client for engine %v", engine.Name)
+	}
+
+	return engineapi.GetCompatibleClient(engine, engineCliClient, vc.ds, nil, vc.proxyConnCounter)
+}
```
Comment (innobead) [inline metrics_collector/volume_collector.go:218]: nit: See several places using this function, so it's time to move to a common place.

**Review by innobead — CHANGES_REQUESTED** [review]
In general, LGTM. Just two comments.

File: engineapi/proxy_metrics.go (hunk 3 above)
Comment (derekbit) [reply engineapi/proxy_metrics.go:12]: It's a common pattern in longhorn-manager.

> Why not just use the type from instance manager?

Yeah, using the same type is a idea way. We will refactor them in the future.

**Review by innobead — APPROVED** [review]

**Comment by innobead** [comment]
@Mergifyio backport v1.4.x

**Pull request merged**